### PR TITLE
add ability to push a mode if the current mode is an ancestor

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79143
* #78847
* __->__ #78822

Discussed briefly [here](https://github.com/pytorch/pytorch/pull/78677#discussion_r887431040) this lets a user set a mode whose inner is already set if that inner is the current mode or its an ancestor. This will be necessary for some more complicated use cases of composition and nice for non-lexical scoping

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -1285,6 +1301,7 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                 called += 1
                 return super().__torch_dispatch__(func, types, args, kwargs)
 
+        print(torch._C._get_torch_dispatch_mode())
```
Comment (ezyang) [inline test/test_python_dispatch.py:1304]: kill me

**Review by ezyang — APPROVED** [review]
One thought: should we always allow pushing in this way, or should you explicitly have to say something like `mode.restore()` to trigger this.

**Comment by ezyang** [comment]
one reason I think it might be reasonable to have a separate API is that when you saving mode to use it later it's a pretty different use case than pushing the mode. We could easily explain that you can't reuse a mode and if you want to restore it you should use the restore method in the error message.

File: torch/utils/_python_dispatch.py [hunk 2]
```diff
@@ -143,20 +143,29 @@ class TorchDispatchMode(metaclass=TorchDispatchModeMeta):
...
     def __enter__(self):
-        if hasattr(self, "inner"):
-            raise RuntimeError(f"{self} has already been used as a mode, please create and use a fresh version")
         old = _get_torch_dispatch_mode()
-        self.inner = old
+        if hasattr(self, "inner"):
+            assert hasattr(self, "ancestors")
+            if old is not None and old not in self.ancestors:
+                raise RuntimeError(f"{self} has already been used as a mode and is not valid in the current state, " +
+                                   "because the current mode is not its ancestor. Please use a fresh version")
```
Comment (zou3519) [inline torch/utils/_python_dispatch.py:157]: We've got these new error messages, but I don't see a test testing them. Should we add some?

**Review by zou3519 — APPROVED** [review]

**Comment by samdow** [comment]
Moved to `restore` so now a user will have to do

`with x.restore()` or `with restore(x):` if `x` has already been used

**Comment by samdow** [comment]
pytorchbot merge -a

**Comment by samdow** [comment]
pytorchbot merge

**Pull request closed without merging**